compiler/rustc_trait_selection/src/traits/normalize.rs

```diff
@@ -169,6 +170,39 @@ impl<'a, 'b, 'tcx> AssocTypeNormalizer<'a, 'b, 'tcx> {
 
         if !needs_normalization(self.selcx.infcx, &value) { value } else { value.fold_with(self) }
     }
+
+    fn normalize_opaque_ty(&mut self, ty: Ty<'tcx>, data: ty::AliasTy<'tcx>) -> Ty<'tcx> {
+        let recursion_limit = self.cx().recursion_limit();
+        if !recursion_limit.value_within_limit(self.depth) {
+            self.selcx.infcx.err_ctxt().report_overflow_error(
+                OverflowCause::DeeplyNormalize(data.into()),
+                self.cause.span,
+                true,
+                |_| {},
+            );
+        }
+
+        let args = data.args.fold_with(self);
+        let generic_ty = self.cx().type_of(data.def_id);
+        let mut concrete_ty = generic_ty.instantiate(self.cx(), args);
+
+        if concrete_ty == ty {
```

Comment:
This recursiveness check sucks, but it prevents bad error messages.

Reply:
could we move this check to `type_of_opaque`?